src/fstab-generator/fstab-generator.c

```diff
@@ -763,21 +763,52 @@ static int parse_fstab(bool initrd) {
                         flags |= rw_only * MOUNT_RW_ONLY |
                                  automount * MOUNT_AUTOMOUNT;
 
+                        is_sysroot = path_equal(where, "/sysroot");
+                        /* See comment from add_sysroot_usr_mount about the need for extra indirection
+                         * in case /usr needs to be mounted in order for the root fs to be synthesized
+                         * based on configuration included in /usr/, e.g. systemd-repart. */
+                        is_sysroot_usr = path_equal(where, "/sysroot/usr");
+
                         const char *target_unit =
                                 initrd ?               SPECIAL_INITRD_FS_TARGET :
+                                is_sysroot ?           SPECIAL_INITRD_ROOT_FS_TARGET :
+                                is_sysroot_usr ?       SPECIAL_INITRD_USR_FS_TARGET :
                                 mount_is_network(me) ? SPECIAL_REMOTE_FS_TARGET :
                                                        SPECIAL_LOCAL_FS_TARGET;
 
+                        if (is_sysroot && is_device_path(what)) {
+                                r = generator_write_initrd_root_device_deps(arg_dest, what);
+                                if (r < 0)
+                                        return r;
+                        }
+
                         k = add_mount(fstab,
                                       arg_dest,
                                       what,
-                                      canonical_where ?: where,
+                                      canonical_where ?: is_sysroot_usr ? "/sysusr/usr" : where,
                                       canonical_where ? where: NULL,
                                       me->mnt_type,
                                       me->mnt_opts,
                                       me->mnt_passno,
                                       flags,
                                       target_unit);
+
+                        if (is_sysroot_usr && k >= 0) {
+                                log_debug("Synthesizing fstab entry what=/sysusr/usr where=/sysroot/usr opts=bind");
+
+                                r = add_mount(fstab,
+                                              arg_dest,
+                                              "/sysusr/usr",
+                                              canonical_where ?: where,
+                                              canonical_where ? where: NULL,
```

Comment:
why use canonical_where/where here?

for the sake of readability, we already established that where is "/sysroot/usr" here, hence I think it would be wise to spell that out, just like we do belo in add_sysroot_usr_mount

Reply:
in fact, I think it would be good if we'd unify this call to add_mount() here and down below, i.e. add a helper add_sysusr_sysroot_usr_bind_mount() or so, that adds the bind mount, and is called from both places. After all, afaics they only differ in a single argument: the source.

Reply:
I originally had that but you're right it would be clearer to just spell it out. I've also factored it out to a helper function like you suggested. Thank you!